---
title: "Confirm protein expression change in HPA"
output:
  html_notebook:
    code_folding: show
editor_options:
  chunk_output_type: console
---
# {.tabset .tabset-fade .tabset-pills}

## Setup
### Load libraries
```{r}
library(here)
library(tidyverse)
library(tidylog)
library(conflicted)
conflict_prefer("filter", "tidylog")
conflict_prefer("drop_na", "tidylog")
conflict_prefer("inner_join", "tidylog")
conflict_prefer("count", "tidylog")
conflict_prefer("intersect", "dplyr")
conflict_prefer("mutate", "tidylog")
conflict_prefer("select", "tidylog")
conflict_prefer("distinct", "tidylog")
conflict_prefer("inner_join", "tidylog")
```

## Download normal and pathology datasets downloaded from HPA
```{r}
## Normal
url <- "https://www.proteinatlas.org/download/normal_tissue.tsv.zip"
file <- here("data", "hpa-normal.tsv.zip")

if(!file.exists(file)){
  download.file(url = url, destfile = file)
}

HPA_normal <- read_tsv(file) %>%
  janitor::clean_names()

## Pathology
url <- "https://www.proteinatlas.org/download/pathology.tsv.zip"
file <- here("data", "hpa-cancer.tsv.zip")

if(!file.exists(file)){
download.file(url = url, destfile = file)
}
HPA_pathology <- read_tsv(file) %>%
  janitor::clean_names()
```

## Filter the HPA by tissue, cell type and the up- or down-regulated genes
```{r}
filterNormal <- function(genes, normal_data){
  normal_data %>%
  filter(tissue %in% c("rectum", "colon", "prostate", "breast", "lung")) %>%
  filter(cell_type %in% c("glandular cells", "pneumocytes")) %>%
  filter(gene %in% genes)
}

filterCancer <- function(genes, cancer_data){
  cancer_data %>%
  select(gene:not_detected) %>% 
  mutate(cancer = str_remove(cancer, pattern = " cancer")) %>% 
  filter(cancer %in% c("colorectal",
                       "prostate",
                       "breast",
                       "lung")) %>% 
  filter(gene %in% genes) %>%
  drop_na(high:not_detected) %>%
  mutate(n_samples = rowSums(select(., high:not_detected)))
}

## Filter the up-regulated genes
degs_4vs2_up <- degs_4vs2 %>% filter(log2FoldChange >= 1)
normal_data_for_up <- filterNormal(genes = degs_4vs2_up$gene_id, normal_data = HPA_normal)
cancer_data_for_up <- filterCancer(genes = degs_4vs2_up$gene_id, cancer_data = HPA_pathology)
## How many up-regulated genes have normal and pathology data in HPA?
intersect(normal_data_for_up$gene, cancer_data_for_up$gene) %>% length() # 446 of 639

## Filter the down-regulated genes
degs_4vs2_down <- degs_4vs2 %>% filter(log2FoldChange <= -1)
normal_data_for_down <- filterNormal(genes = degs_4vs2_down$gene_id, normal_data = HPA_normal)
cancer_data_for_down <- filterCancer(genes = degs_4vs2_down$gene_id, cancer_data = HPA_pathology)
## How many of the down-regulated genes have normal and pathology data in HPA?
intersect(normal_data_for_down$gene, cancer_data_for_down$gene) %>% length()# 949 of 1329
```

## Check for differences in protein abundance between normal tissue and cancer
### Define the comparison functions
```{r}
checkIfUpregulated <- function(tissue_type,
                               cancer_type,
                               normal_data = normal_data_for_up,
                               cancer_data = cancer_data_for_up)
  {
  inner_join(normal_data, cancer_data, by = "gene") %>%
    select(gene, gene_name = gene_name.x, tissue, cancer, level, high:not_detected, n_samples) %>% 
    mutate(samples_up = case_when(level == "High" ~ 0,
                                   level == "Medium" ~ as.double(high),
                                   level == "Low" ~ as.double(high+medium),
                                   level == "Not detected" ~ as.double(high+medium+low))) %>% 
    mutate(samples_down = case_when(level == "High" ~ as.double(medium+low+not_detected),
                                   level == "Medium" ~ as.double(low+not_detected),
                                   level == "Low" ~ as.double(not_detected),
                                   level == "Not detected" ~ 0)) %>% 
    mutate(samples_noChange = case_when(level == "High" ~ as.double(high),
                                   level == "Medium" ~ as.double(medium),
                                   level == "Low" ~ as.double(low),
                                   level == "Not detected" ~ as.double(not_detected))) %>% 
    filter(samples_up >= 0.5 * n_samples) %>% 
    filter(tissue == tissue_type) %>% 
    filter(cancer == cancer_type) %>%
    select(gene, gene_name) %>% 
    arrange(gene_name)
}

checkIfDownregulated <- function(tissue_type,
                                 cancer_type,
                                 normal_data = normal_data_for_down, 
                                 cancer_data = cancer_data_for_down)
  {
  inner_join(normal_data, cancer_data, by = "gene") %>%
    select(gene, gene_name = gene_name.x, tissue, cancer, level, high:not_detected, n_samples) %>% 
    mutate(samples_down = case_when(level == "High" ~ 0,
                                   level == "Medium" ~ as.double(high),
                                   level == "Low" ~ as.double(high+medium),
                                   level == "Not detected" ~ as.double(high+medium+low))) %>% 
    mutate(samples_down = case_when(level == "High" ~ as.double(medium+low+not_detected),
                                   level == "Medium" ~ as.double(low+not_detected),
                                   level == "Low" ~ as.double(not_detected),
                                   level == "Not detected" ~ 0)) %>% 
    mutate(samples_noChange = case_when(level == "High" ~ as.double(high),
                                   level == "Medium" ~ as.double(medium),
                                   level == "Low" ~ as.double(low),
                                   level == "Not detected" ~ as.double(not_detected))) %>% 
    filter(samples_down >= 0.5 * n_samples) %>% 
    filter(tissue == tissue_type) %>% 
    filter(cancer == cancer_type) %>% 
    select(gene, gene_name) %>% 
    arrange(gene_name)
}
```

### Check for up-regulation
```{r}
breast_up <- checkIfUpregulated(tissue_type = "breast", cancer_type = "breast") # 64
colon_up <- checkIfUpregulated(tissue_type = "colon", cancer_type = "colorectal") # 66
rectum_up <- checkIfUpregulated(tissue_type = "rectum", cancer_type = "colorectal") # 64
colorectal_up <- semi_join(colon_up, rectum_up) %>% arrange(gene_name) # 54
lung_up <- checkIfUpregulated(tissue_type = "lung", cancer_type = "lung") # 104
prostate_up <- checkIfUpregulated(tissue_type = "prostate", cancer_type = "prostate") # 72
```

### Create a Venn diagram of up-regulated genes
```{r warning=F}
venn_up <- gplots::venn(
  list("BRCA" = breast_up$gene_name,
       "COADREAD" = colorectal_up$gene_name,
       "LUAD" = lung_up$gene_name,
       "PRAD" = prostate_up$gene_name))

common_up <- attr(venn_up, "intersections")$`BRCA:COADREAD:LUAD:PRAD`
  # 13 proteins up-regulated in all 4 cancers
bind_rows(breast_up, colorectal_up, lung_up, prostate_up) %>% distinct() %>% count()
  # there are 178 uniqe up-regulated proteins
```

### Check for down-regulation
```{r}
breast_down <- checkIfDownregulated(tissue_type = "breast", cancer_type = "breast") # 239
colon_down <- checkIfDownregulated(tissue_type = "colon", cancer_type = "colorectal") # 305
rectum_down <- checkIfDownregulated(tissue_type = "rectum", cancer_type = "colorectal") # 319
colorectal_down <- semi_join(colon_down, rectum_down) %>% arrange(gene_name)# 258
lung_down <- checkIfDownregulated(tissue_type = "lung", cancer_type = "lung") # 218
prostate_down <- checkIfDownregulated(tissue_type = "prostate", cancer_type = "prostate") # 265
```

### Create a Venn diagram of down-regulated genes
```{r warning=F}
venn_down <- gplots::venn(
  list("BRCA" = breast_down$gene_name,
       "COADREAD" = colorectal_down$gene_name,
       "LUAD" = lung_down$gene_name,
       "PRAD" = prostate_down$gene_name))

common_down <- attr(venn_down, "intersections")$`BRCA:COADREAD:LUAD:PRAD`
  # 29 proteins down-regulated in all 4 cancers
bind_rows(breast_down, colorectal_down, lung_down, prostate_down) %>% distinct() %>% count()
  # there are 552 uniqe down-regulated proteins 
```

### Create Venn diagram for all DEGs
```{r warning=F}
venn_all_degs <- gplots::venn(
  list("BRCA" = degs_breast$gene_name,
       "COADREAD" = degs_colorectal$gene_name,
       "LUAD" = degs_lung$gene_name,
       "PRAD" = degs_prostate$gene_name))

common_degs <- attr(venn_all_degs, "intersections")$`BRCA:COADREAD:LUAD:PRAD`
```

## Place the 3 Venn diagrams on one figure
```{r}
grDevices::pdf(file = here("output", "degs-after-HPA_venn.pdf"))
graphics::par(mfrow = c(2,2), cex = 0.7, oma = c(0, 0, 4, 0))
graphics::par(mar = c(0,0,2,0))
graphics::plot(venn_up)
graphics::title(main = "up-regulated")
graphics::par(mar = c(0,0,2,0))
graphics::plot(venn_down)
graphics::title(main = "down-regulated")
graphics::par(mar = c(0,0,1,0), fig = c(0.25,0.75,0,0.5), new = T)
graphics::plot(venn_all_degs)
graphics::title(main = "all DEGs")
graphics::mtext("Overlap between genes differentially expressed at a transcript and protein level", outer = T, cex = 1, padj = -2)
grDevices::dev.off()
```

## Check if any of the genes are in the Cancer Gene Census
```{r}
cgc <- read_tsv(here("data", "Census_allSun Jan 19 20_33_11 2020.tsv"))

cgc %>% filter_at("Gene Symbol", any_vars(. %in% common_degs)) # CD74, PBX1
```

## Save results to .xlsx file and clean up
```{r}
mapping_ensg2HGNC <- bind_rows(breast_up, breast_down, colorectal_up, colorectal_down, lung_up, lung_down, prostate_up, prostate_down) %>% distinct()

has_HPA_data <- c(intersect(normal_data_for_up$gene, cancer_data_for_up$gene), intersect(normal_data_for_down$gene, cancer_data_for_down$gene))

degs_4vs2 %>%
  inner_join(mapping_ensg2HGNC, by = c("gene_id" = "gene")) %>% 
  select(-(3:7)) %>% 
  mutate("brca_up" = case_when(gene_id %in% breast_up$gene ~ "Yes")) %>%
  mutate("coadread_up" = case_when(gene_id %in% colorectal_up$gene ~ "Yes")) %>%
  mutate("luad_up" = case_when(gene_id %in% lung_up$gene ~ "Yes")) %>%
  mutate("prad_up" = case_when(gene_id %in% prostate_up$gene ~ "Yes")) %>%
  mutate("brca_down" = case_when(gene_id %in% breast_down$gene ~ "Yes")) %>%
  mutate("coadread_down" = case_when(gene_id %in% colorectal_down$gene ~ "Yes")) %>%
  mutate("luad_down" = case_when(gene_id %in% lung_down$gene ~ "Yes")) %>%
  mutate("prad_down" = case_when(gene_id %in% prostate_down$gene ~ "Yes")) %>%
  mutate("n_cancers_w_change" = rowSums(.[4:11] == "Yes", na.rm = T)) %>%  # how many cancers have this protein altered
  mutate("in_CGC" = case_when(gene_name %in% cgc$`Gene Symbol` ~ "Yes")) %>% # check if in CGC
  select("ensg_id" = gene_id,
         "gene_symbol" = gene_name,
         "transcript_l2fc" = log2FoldChange,
         n_cancers_w_change,
         in_CGC,
         everything()) #%>% 
  degs_4vs2_after_HPA %>% 
  write_tsv(here("output", "degs_4vs2_after_HPA.tsv"))
```

## Save the DEGs to a .tsv file
```{r}
rowr::cbind.fill(breast_up$gene_name,
                 colorectal_up$gene_name,
                 lung_up$gene_name,
                 prostate_up$gene_name,
                 common_up,
                 breast_down$gene_name,
                 colorectal_down$gene_name,
                 lung_down$gene_name,
                 prostate_down$gene_name,
                 common_down,
                 fill = "") %>%
  setNames(str_c(c("brca", "coadread", "luad", "prad", "common"),
                 c(rep("_up", 5), rep("_down", 5)))) %>%
  write_tsv(here("output", "4vs2_DEGs_after_HPA.tsv"))
```

## Clean up
```{r}
varhandle::rm.all.but(keep = c("all_degs_HPA"), keep_functions = F)
varhandle::save.var(path = here("analysis"), varlist = "all_degs_HPA")
```

